### PR TITLE
Absorption

### PR DESCRIPTION
**Pull request by SelfMadeSystem**
# Problem

See #557

Will be necessary to implement #556 and #558

# Objective

Implement absorption for ALL living entities.

# Solution

Make a special case for LivingEntity for Absorption since it's untracked.

# Playground

<details>
<summary>Playground</summary>

```rs
use valence::client::despawn_disconnected_clients;
use valence::entity::living::Absorption;
use valence::log::LogPlugin;
use valence::network::ConnectionMode;
use valence::prelude::*;

#[allow(unused_imports)]
use crate::extras::*;

const SPAWN_Y: i32 = 64;

pub fn build_app(app: &mut App) {
    app.insert_resource(NetworkSettings {
        connection_mode: ConnectionMode::Offline,
        ..Default::default()
    })
    .add_plugins(DefaultPlugins.build().disable::<LogPlugin>())
    .add_systems(Startup, setup)
    .add_systems(EventLoopUpdate, toggle_absorption_on_sneak)
    .add_systems(Update, (init_clients, despawn_disconnected_clients))
    .run();
}

fn setup(
    mut commands: Commands,
    server: Res<Server>,
    biomes: Res<BiomeRegistry>,
    dimensions: Res<DimensionTypeRegistry>,
) {
    let mut layer = LayerBundle::new(ident!("overworld"), &dimensions, &biomes, &server);

    for z in -5..5 {
        for x in -5..5 {
            layer.chunk.insert_chunk([x, z], UnloadedChunk::new());
        }
    }

    for z in -25..25 {
        for x in -25..25 {
            layer
                .chunk
                .set_block([x, SPAWN_Y, z], BlockState::GRASS_BLOCK);
        }
    }

    commands.spawn(layer);
}

fn init_clients(
    mut clients: Query<
        (
            &mut EntityLayerId,
            &mut VisibleChunkLayer,
            &mut VisibleEntityLayers,
            &mut Position,
            &mut GameMode,
        ),
        Added<Client>,
    >,
    layers: Query<Entity, (With<ChunkLayer>, With<EntityLayer>)>,
) {
    for (
        mut layer_id,
        mut visible_chunk_layer,
        mut visible_entity_layers,
        mut pos,
        mut game_mode,
    ) in &mut clients
    {
        let layer = layers.single();

        layer_id.0 = layer;
        visible_chunk_layer.0 = layer;
        visible_entity_layers.0.insert(layer);
        pos.set([0.0, SPAWN_Y as f64 + 1.0, 0.0]);
        *game_mode = GameMode::Survival;
    }
}

pub fn toggle_absorption_on_sneak(
    mut clients: Query<&mut Absorption>,
    mut events: EventReader<SneakEvent>,
) {
    for event in events.iter() {
        if event.state == SneakState::Start {
            if let Ok(mut absorption) = clients.get_mut(event.client) {
                absorption.0 = if absorption.0 == 0.0 { 4.0 } else { 0.0 };
            }
        }
    }
}
```
</details>

### CODE REVIEW

File: crates/valence_entity/build.rs [hunk 1]
```diff
@@ -506,6 +524,22 @@ fn build() -> anyhow::Result<TokenStream> {
         }]);
     }
 
+    systems.extend([quote! {
+        #[allow(clippy::needless_borrow)]
+        #[allow(clippy::suspicious_else_formatting)]
```
Comment (dyc3) [inline crates/valence_entity/build.rs:529]: do we need to ignore these lints?

File: crates/valence_entity/build.rs [hunk 2]
```diff
@@ -506,6 +524,22 @@ fn build() -> anyhow::Result<TokenStream> {
...
+        /// Special case for `living::Absorption`.
+        /// Updates the `AbsorptionAmount` component of the player entity.
+        fn update_living_and_player_absorption(
+            mut query: Query<(&living::Absorption, &mut player::AbsorptionAmount), Changed<living::Absorption>>
```
Comment (dyc3) [inline crates/valence_entity/build.rs:533]: What's the difference between `living::Absorption` and `player::AbsorptionAmount`? Feels like the same thing.

File: crates/valence_entity/build.rs [hunk 3]
```diff
@@ -506,6 +524,22 @@ fn build() -> anyhow::Result<TokenStream> {
...
+        ) {
+            for (living_absorption, mut player_absorption) in &mut query {
```
Comment (dyc3) [inline crates/valence_entity/build.rs:535]: nit:
```suggestion
            for (living_absorption, mut player_absorption) in query.iter_mut() {
```

**Review by dyc3 — CHANGES_REQUESTED** [review]

File: crates/valence_entity/build.rs (hunk 2 above)
Comment (SelfMadeSystem) [reply crates/valence_entity/build.rs:533]: `player::AbsorptionAmount` is a **tracked** field (so updated via the `EntityTrackerUpdateS2c` packet), but `living::Absorption` is NOT tracked. `player::AbsorptionAmount` is only tracked to update the client on their own absorption level. Furthermore, `player::AbsorptionAmount` is **only** for the client, not for any other entity.

**Review by dyc3 — APPROVED** [review]

**Pull request merged**